pandas/parser.pyx

```diff
@@ -217,6 +221,24 @@ cdef extern from "parser/tokenizer.h":
     int to_boolean(const char *item, uint8_t *val) nogil
 
 
+# XXX
+# this is a hack - in order to make the inference
+# functions generic (converting either data directly
+# from the parser or from a passed in hash table)
+# we add an "optional" parameter via fused type, that can either
+# be the hash table to parse, or an integer, which is used
+# as a sentinel to specialize the function for reading
+# from the parser.
+
+# This is to avoid duplicating a bunch of code or
```

Comment:
if I put it in the parser state then checks like [this](https://github.com/pydata/pandas/pull/13406/files#diff-10cd69d969114cd15ba284e3e5da6e22R1720) would get moved to runtime, which I was trying to avoid, though to be fair I haven't profiled to see what it costs.